### PR TITLE
Feat: missing fuel via emission category validation

### PR DESCRIPTION
**Pull request by dleard**
Function validates that if an emission category contains an emission reported with annual_co2e > 0 then a fuel should be reported (with quantity > 0) in that category as well.
https://youtrack.button.is/issue/GGIRCS-2314

### CODE REVIEW

**Comment by dleard** [comment]
Validation message on the summary page:
missingfuel

File: schema/deploy/application_validation_functions/emission_category_missing_fuel.sql [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    select distinct(display_name) as reported_via_fuel from ggircs_portal.emission_category ec
+    join fuel_data
+      on fuel_data.emission_category_id = ec.id
+      and fuel_data.quantity > 0
+  ) select (select count(*) from emissions where reported_via_emission not in (select reported_via_fuel from fuels)) = 0;
+
+$$ language sql stable;
+
+grant execute on function ggircs_portal.emission_category_missing_fuel to ciip_administrator, ciip_analyst, ciip_industry_user, ciip_guest;
+
```
Comment (pbastia) [inline schema/deploy/application_validation_functions/emission_category_missing_fuel.sql:28]: do we need `ciip_guest` for this function?

Comment (dleard) [reply schema/deploy/application_validation_functions/emission_category_missing_fuel.sql:28]: nope!

File: schema/deploy/application_validation_functions/emission_category_missing_fuel.sql [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    where ed.annual_co2e > 0
+  ), fuels as (
+    with fuel_data as (
+      select * from ggircs_portal.application_revision_fuel_form_data(app_revision)
+    )
+    select distinct(display_name) as reported_via_fuel from ggircs_portal.emission_category ec
+    join fuel_data
+      on fuel_data.emission_category_id = ec.id
+      and fuel_data.quantity > 0
+  ) select (select count(*) from emissions where reported_via_emission not in (select reported_via_fuel from fuels)) = 0;
```
Comment (pbastia) [inline schema/deploy/application_validation_functions/emission_category_missing_fuel.sql:23]: For readability I'd suggest making the `with fuel_data` not nested in the second `with`

```sql
  with emissions as (
    select distinct(source_type_name) as reported_via_emission
    from ggircs_portal.application_revision_emission_form_data(app_revision) ed
    where ed.annual_co2e > 0
  ), 
  fuel_data as (
    select * from ggircs_portal.application_revision_fuel_form_data(app_revision)
  ),
  fuels as (
    select distinct(display_name) as reported_via_fuel from ggircs_portal.emission_category ec
    join fuel_data
  ...
```

File: schema/test/unit/application_validation_functions/emission_category_missing_fuel_test.sql [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    select ggircs_portal.emission_category_missing_fuel((select * from record))),
+  false,
+  'Function returns false when there are emission categories with an emission reported, but no corresponding fuel reported'
+);
+
+select is(
+  (with record as (select row(application_revision.*)::ggircs_portal.application_revision from ggircs_portal.application_revision where application_id=2 and version_number=1)
+    select ggircs_portal.emission_category_missing_fuel((select * from record))),
+  true,
+  'Function returns false when there are emission categories with an emission reported, but no corresponding fuel reported'
```
Comment (pbastia) [inline schema/test/unit/application_validation_functions/emission_category_missing_fuel_test.sql:96]:
```suggestion
'Function returns true when ...'
```

**Review by pbastia — CHANGES_REQUESTED** [review]
just a couple of style things. Very nice!

**Review by pbastia — APPROVED** [review]

**Pull request merged**